distribution/xfer/download.go

```diff
@@ -93,37 +104,64 @@ func (ldm *LayerDownloadManager) Download(ctx context.Context, initialRootFS ima
 		topLayer       layer.Layer
 		topDownload    *downloadTransfer
 		watcher        *Watcher
-		missingLayer   bool
 		transferKey    = ""
 		downloadsByKey = make(map[string]*downloadTransfer)
+		cacheDirs      = map[string]string{}
+		success        = false
 	)
 
+	defer func() {
+		if success {
+			for key := range cacheDirs {
+				err := ldm.cacheDirManager.DeleteDir(key)
+				if err != nil {
+					logrus.Warnf("failed to delete cache dir: %s", err)
+				}
+			}
+		} else {
+			for key := range cacheDirs {
+				ldm.cacheDirManager.ReleaseDir(key)
+			}
+		}
+	}()
+
+	// Look for a rootFS/layer we can resume from before trying to download layers.
+	// This loop goes up to resumeLevel layers
 	rootFS := initialRootFS
+	resumeLevel := 0
+	transferKey = ""
 	for _, descriptor := range layers {
+		diffID, err := descriptor.DiffID()
+		if err != nil {
+			continue
```

Comment:
I think this should be a `break`. It would not be correct to continue appending layers to `rootFS` after having skipped one.